src/EditorFeatures/Core/LanguageServer/Handlers/CodeActions/CodeActionsCache.cs

```diff
@@ -2,22 +2,38 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Composition;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Completion;
 using Microsoft.CodeAnalysis.UnifiedSuggestions;
 using Roslyn.Utilities;
 using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions
 {
+    [ExportRoslynLspServiceFactory(typeof(CodeActionsCache)), Shared]
+    internal class CodeActionsCacheFactory : ILspServiceFactory
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CodeActionsCacheFactory()
+        {
+        }
+
+        public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind) => new CodeActionsCache();
```

Comment:
If we return a new cache each time this is called and `CodeActionResolveHandler.cs` doesn't hold it then I assume that the Context object is ensuring that only one of these is created for the whole process?

My paranoia makes me think it might be good to add a quick check here to see that this is only called once if that's the case.

Reply:
To clarify - `CreateILspService` is called once per server.  Then everything in that server is handed a `LspServices` instance that holds onto the appropriate lsp services for that server.

So on each GetService call it returns the instance created for that server.  I have unit tests that verify this, an assert here would be kind of hard to write.